components/raftstore-v2/src/operation/ready/snapshot.rs

```diff
@@ -318,28 +314,31 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
         true
     }
 
-    /// Cancel generating snapshot.
-    pub fn cancel_generating_snap(&self, compact_to: Option<u64>) {
-        let mut snap_state = self.snap_state_mut();
-        let SnapState::Generating {
-           ref canceled,
-           ref index,
-        } = *snap_state else { return };
-
-        if let Some(idx) = compact_to {
-            let snap_index = index.load(Ordering::SeqCst);
-            if snap_index == 0 || idx <= snap_index + 1 {
-                return;
+    pub fn cancel_generating_snap(&self, to: Option<u64>, compact_to: Option<u64>) {
+        if let Some(id) = to {
+            let mut states = self.snap_states.borrow_mut();
+            if let Some(state) = states.get(&id) {
+                let SnapState::Generating {
+                    ref canceled,
+                    ref index,
+                } = *state else { return };
+                if let Some(idx) = compact_to {
+                    let snap_index = index.load(Ordering::SeqCst);
+                    if snap_index == 0 || idx <= snap_index + 1 {
+                        return;
+                    }
+                }
+                info!(
+                    self.logger(),
+                    "snapshot is canceled";
+                    "compact_to" => compact_to,
+                );
+                states.remove(&id);
             }
+        } else {
+            self.snap_states.borrow_mut().clear();
         }
-        canceled.store(true, Ordering::SeqCst);
-        *snap_state = SnapState::Relax;
         self.gen_snap_task_mut().take();
```

Comment:
I think the `gen_snap_task`  doesn't specify the `to_peer_id`, so it will be clear if the cancel has been called. It will generate one task in next snapshot.

Reply:
The task will be taken no matter matched the to_peer_id, it will generate new task in next snapshot. The `gen_snap_task ` shouldn't be associated with one `to_peer_id`.

Reply:
It will be canceled , I will verify the peer_id before cancel.